pkg/service/oidc/service.go

```diff
@@ -0,0 +1,296 @@
+package oidc
+
+import (
+	"context"
+	"crypto/rand"
+	"net/http"
+	"time"
+
+	didsdk "github.com/TBD54566975/ssi-sdk/did"
+	"github.com/TBD54566975/ssi-sdk/oidc/issuance"
+	"github.com/TBD54566975/ssi-sdk/util"
+	"github.com/google/uuid"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pkg/errors"
+	"github.com/pquerna/otp"
+	"github.com/pquerna/otp/totp"
+	"github.com/sirupsen/logrus"
+	credint "github.com/tbd54566975/ssi-service/internal/credential"
+	"github.com/tbd54566975/ssi-service/internal/did"
+	"github.com/tbd54566975/ssi-service/internal/keyaccess"
+	"github.com/tbd54566975/ssi-service/pkg/server/framework"
+	"github.com/tbd54566975/ssi-service/pkg/service/credential"
+	svcframework "github.com/tbd54566975/ssi-service/pkg/service/framework"
+	"github.com/tbd54566975/ssi-service/pkg/service/oidc/model"
+)
+
+const DefaultCNonceExpiration = 5 * time.Second
+
+type Service struct {
+	// key to use for generating nonce values.
+	key             *otp.Key
+	cNonceExpiresIn time.Duration
+
+	resolver    didsdk.Resolver
+	credService *credential.Service
+}
+
+var _ svcframework.Service = (*Service)(nil)
+
+func (s Service) Type() svcframework.Type {
+	return svcframework.OIDC
+}
+
+func (s Service) Status() svcframework.Status {
+	return svcframework.Status{Status: svcframework.StatusReady}
+}
+
+type Option func(*Service)
+
+func WithCNonceExpiresIn(d time.Duration) Option {
+	return func(service *Service) {
+		service.cNonceExpiresIn = d
+	}
+}
+
+func WithOTPKey(k *otp.Key) Option {
+	return func(service *Service) {
+		service.key = k
+	}
+}
+
+func NewOIDCService(didResolver didsdk.Resolver, service *credential.Service, serviceKey [32]byte, opts ...Option) *Service {
+	s := &Service{
+		cNonceExpiresIn: DefaultCNonceExpiration,
+		resolver:        didResolver,
+		credService:     service,
+	}
+	for _, o := range opts {
+		o(s)
+	}
+	if s.key == nil {
+		key, err := totp.Generate(totp.GenerateOpts{
+			Issuer:      "ssi-service",
+			AccountName: "internal",
+			Period:      uint(s.cNonceExpiresIn / time.Second),
+			SecretSize:  32,
+			Digits:      16,
+			Rand:        rand.Reader,
+			Algorithm:   otp.AlgorithmSHA512,
+			Secret:      serviceKey[:],
+		})
+		if err != nil {
+			panic(err)
+		}
+		WithOTPKey(key)(s)
+	}
+
+	return s
+}
+
+func (s Service) CredentialEndpoint(ctx context.Context, credRequest *model.CredentialRequest) (*model.CredentialResponse, error) {
+	if credRequest.Format == "" {
+		return nil, framework.NewRequestErrorMsg("invalid_request", http.StatusBadRequest)
```

Comment:
would consider creating a const block of errors ...

`const (
 InvalidRequest OurErrorType = "invalid_request"
 ...
)`